container/container.go

```diff
@@ -2,19 +2,91 @@ package container
 
 import (
 	"context"
+	"time"
 
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
 	docker "github.com/docker/docker/client"
 )
 
-// FindContainer returns a docker container.
-func FindContainer(namespace []string) (types.ContainerJSON, error) {
-	client, err := Client()
+// Container provides high level interactions with Docker API for MESG.
+type Container struct {
+	// client is a Docker client.
+	client docker.CommonAPIClient
+
+	// callTimeout is the timeout value for Docker API calls.
+	callTimeout time.Duration
+}
+
+// Option is a configuration func for Container.
+type Option func(*Container)
+
+// New creates a new Container with given options.
+func New(options ...Option) (*Container, error) {
+	c := &Container{
+		callTimeout: time.Second * 10,
+	}
+	for _, option := range options {
+		option(c)
+	}
+	var err error
+	if c.client == nil {
+		c.client, err = docker.NewClientWithOpts(docker.FromEnv)
+		if err != nil {
+			return c, err
+		}
+	}
+	c.negotiateAPIVersion()
+	if err := c.createSwarmIfNeeded(); err != nil {
+		return c, err
+	}
+	return c, c.createSharedNetworkIfNeeded()
+}
+
+// ClientOption receives a client which will be used to interact with Docker API.
+func ClientOption(client docker.CommonAPIClient) Option {
+	return func(c *Container) {
+		c.client = client
+	}
+}
+
+// TimeoutOption receives d which will be set as a timeout value for Docker API calls.
+func TimeoutOption(d time.Duration) Option {
+	return func(c *Container) {
+		c.callTimeout = d
+	}
+}
+
+func (c *Container) negotiateAPIVersion() {
+	ctx, cancel := context.WithTimeout(context.Background(), c.callTimeout)
+	defer cancel()
+	c.client.NegotiateAPIVersion(ctx)
+}
+
+func (c *Container) createSwarmIfNeeded() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.callTimeout)
+	defer cancel()
+	info, err := c.client.Info(ctx)
 	if err != nil {
-		return types.ContainerJSON{}, err
+		return err
+	}
+	if info.Swarm.NodeID != "" {
+		return nil
 	}
-	containers, err := client.ContainerList(context.Background(), types.ContainerListOptions{
+	ctx, cancel = context.WithTimeout(context.Background(), c.callTimeout)
+	defer cancel()
```

Comment:
is it ok to assign `cancel` multiple time and call it using `defer`?
Is `defer` will really execute the 2 cancel or the last cancel 2 times?

Reply:
yes it is. please see: https://play.golang.org/p/UeNDuSXbZz7